packages/contracts-bedrock/contracts/L1/L2OutputOracleV2.sol

```diff
@@ -0,0 +1,331 @@
+            "L2OutputOracleV2: minimum proposal cost not provided"
+        );
+
+        require(
+            l2Outputs[_l2BlockNumber].timestamp == 0,
+            "L2OutputOracleV2: output already proposed"
+        );
+
+        require(
+            computeL2Timestamp(_l2BlockNumber) < block.timestamp,
+            "L2OutputOracleV2: cannot propose L2 output in the future"
+        );
+
+        require(
+            _outputRoot != bytes32(0),
+            "L2OutputOracleV2: L2 output proposal cannot be the zero hash"
+        );
+
+        if (_l1BlockHash != bytes32(0)) {
+            // This check allows the proposer to propose an output based on a given L1 block,
+            // without fear that it will be reorged out.
+            // It will also revert if the blockheight provided is more than 256 blocks behind the
+            // chain tip (as the hash will return as zero). This does open the door to a griefing
+            // attack in which the proposer's submission is censored until the block is no longer
+            // retrievable, if the proposer is experiencing this attack it can simply leave out the
+            // blockhash value, and delay submission until it is confident that the L1 block is
+            // finalized.
+            require(
+                blockhash(_l1BlockNumber) == _l1BlockHash,
+                "L2OutputOracleV2: block hash does not match the hash at the expected height"
+            );
+        }
+
+        // Walk backwards in the doubly linked list
+        uint256 prevBlockNum = highestL2BlockNumber;
+        while (prevBlockNum > _l2BlockNumber) {
+            prevBlockNum = previousBlockNumber[prevBlockNum];
+        }
+        uint256 updatePrevNext = nextBlockNumber[prevBlockNum];
+        if (updatePrevNext != 0) {
+            previousBlockNumber[updatePrevNext] = _l2BlockNumber;
+        }
+        nextBlockNumber[prevBlockNum] = _l2BlockNumber;
+        nextBlockNumber[_l2BlockNumber] = updatePrevNext;
+        previousBlockNumber[_l2BlockNumber] = prevBlockNum;
+
+        // Update the highest L2 block number if necessary.
+        if (_l2BlockNumber > highestL2BlockNumber) {
+            highestL2BlockNumber = _l2BlockNumber;
+        }
+
+        bonds[_l2BlockNumber] = msg.sender;
+
+        l2Outputs[_l2BlockNumber] = Types.OutputProposal({
+            outputRoot: _outputRoot,
+            timestamp: uint128(block.timestamp),
+            l2BlockNumber: uint128(_l2BlockNumber)
+        });
+
+        emit OutputProposed(_outputRoot, _l2BlockNumber, block.timestamp);
+    }
+
+    /**
+     * @notice Returns an output by block number.
+     *
+     * @param _l2BlockNumber L2 block number of the output to return.
+     *
+     * @return The output for the given block number.
+     */
+    function getL2Output(uint256 _l2BlockNumber)
+        external
+        view
+        returns (Types.OutputProposal memory)
+    {
+        return l2Outputs[_l2BlockNumber];
+    }
```

Comment:
Did we consider using inheritance to avoid repeating code that should match between the v1 and v2 L2OO?

Reply:
That's a good point - I think it's still up in the air whether to just replace `L2OutputOracle`'s implementation with  the `L2OutputOracleV2` code or to keep both implementations. If we keep both - a bit of refactoring might be in order.